File: daily-coding-problems/1-arrays/1-3-maximum-subarray-sum/java/solution.java
import java.util.Arrays;

class Solution {
  public static int maxSubArray(int[] input) {
    // Uses Kadane's algorithm
    int maxSeen = 0;
    int maxHere = 0;

    for (int num : input) {
      maxHere += num;
      maxHere = maxHere > 0 ? maxHere : 0;
      maxSeen = maxHere > maxSeen ? maxHere : maxSeen;
    }
    return maxSeen;
  }

  public static int minSubArray(int[] input) {
    int minSeen = 0;
    int minHere = 0;

    for (int num : input) {
      minHere += num;
      minHere = minHere < 0 ? minHere : 0;
      minSeen = minHere < minSeen ? minHere : minSeen;
    }
    return minSeen;
  }

  public static int circularMaxSubAray(int[] input) {
    int min = minSubArray(input);
    int sum = 0;
    for (int num : input) {
      sum += num;
    }
    int circularMaxSubArray = sum - min;
    return Math.max(circularMaxSubArray, maxSubArray(input));
  }

  public static void main(String[] args) {
    int[] ex1 = new int[]{34, -50, 42, 14, -5, 86};
    int[] ex2 = new int[]{-5, -1, -8, -9};

    System.out.println("example1: " + Arrays.toString(ex1));
    System.out.println("example2: " + Arrays.toString(ex2));

    System.out.println("solution1: " + Integer.toString(maxSubArray(ex1)));
    System.out.println("solution2: " + Integer.toString(maxSubArray(ex2)));

    System.out.println("circular1: " + Integer.toString(circularMaxSubAray(ex1)));
    System.out.println("circular2: " + Integer.toString(circularMaxSubAray(ex2)));
  }
}

File: daily-coding-problems/1-arrays/1-3-maximum-subarray-sum/scala/solution.scala
object Solution {
  def maxSubArray(input: Array[Int]): Int = {
    val (max, _) = input.foldLeft[(Int, Int)]((0, 0))((res, num) => {
      var (maxSeen, maxHere) = res
      maxHere += num
      maxHere = math.max(maxHere, 0)
      maxSeen = math.max(maxSeen, maxHere)
      (maxSeen, maxHere)
    })
    return max
  }

  def minSubArray(input: Array[Int]): Int = {
    val (min, _) = input.foldLeft[(Int, Int)]((0, 0))((res, num) => {
      var (minSeen, minHere) = res
      minHere += num
      minHere = math.min(minHere, 0)
      minSeen = math.min(minSeen, minHere)
      (minSeen, minHere)
    })
    return min
  }

  def circularMaxArray(input: Array[Int]): Int = {
    val sum = input.reduceLeft(_ + _)
    val circularSubArray = sum - minSubArray(input)
    return math.max(circularSubArray, maxSubArray(input))
  }

  def main(args: Array[String]): Unit = {
    val ex1 = Array(34, -50, 42, 14, -5, 86)
    val ex2 = Array(-5, -1, -8, -9)

    println("ex1: [" + ex1.mkString(", ") + "]")
    println("ex2: [" + ex2.mkString(", ") + "]")

    println("solution1: " + maxSubArray(ex1))
    println("solution2: " + maxSubArray(ex2))

    println("circular1: " + circularMaxArray(ex1))
    println("circular2: " + circularMaxArray(ex2))
  }
}
